### PR TITLE
Add labkey.webdav

### PR DESCRIPTION
**Pull request by bbimber**
I thought I'd try to start the review around the PR to add support for basic webDav operations.  This PR currently has get/put, which appear to be working, at least for me.  

This PR includes a vignette, not intended for merge, with an example.  I would like to put a basic test case that follows this kind of pattern somewhere.  It seems like the right place for this is SVN, not this git repo?

### CODE REVIEW

**Comment by bbimber** [comment]
@cnathe this is a pretty fleshed out version of things.  it includes webdavTest.R, which is a reasonable test of this new functionality.  I added that file here to show how it works, but it seems like ultimately we should drop that file from the PR and incorporate it into SVN and that test.

**Comment by bbimber** [comment]
@eisascience - this is that PR for Rlabkey and webdav.

File: Rlabkey/R/labkey.webdav.R [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+##
+
+labkey.webdav.get <- function(baseUrl=NULL, folderPath, fileSet="@files", remoteFilePath, localFilePath, overwrite=TRUE)
+{
```
Comment (labkey-klum) [inline Rlabkey/R/labkey.webdav.R]: if you are using default values, they are often more useful if they are at the end of the parameter list, especially if the user is matching parameters by position.

Comment (bbimber) [reply Rlabkey/R/labkey.webdav.R]: sure, good point.  just added that.

File: Rlabkey/R/labkey.webdav.R [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+
+labkey.webdav.pathExists <- function(baseUrl=NULL, folderPath, remoteFilePath, fileSet="@files")
+{
+    baseUrl=labkey.getBaseUrl(baseUrl);
+
+    if (missing(baseUrl) || is.null(baseUrl) || missing(folderPath) || missing(remoteFilePath)) {
+        stop (paste("A value must be specified for each of baseUrl, folderPath, fileSet, and remoteFilePath"));
+    }
+    
+    ret <- labkey.webdav.listDir(baseUrl, folderPath, fileSet, remoteFilePath)
```
Comment (XingY) [inline Rlabkey/R/labkey.webdav.R]: This probably works for R, but it's confusing to non R experts that fileSet and remoteFilePath are swapped in order

File: Rlabkey/R/labkey.webdav.R [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+    if (missing(baseUrl) || is.null(baseUrl) || missing(folderPath) || missing(remoteFilePath)){
+        stop (paste("A value must be specified for each of baseUrl, folderPath, fileSet, and remoteFilePath"))
+    }
+
+    remoteFilePaths <- strsplit(remoteFilePath, "/")[[1]]
+    toCreate <- ""
+    for (folderName in remoteFilePaths) {
+        toCreate <- paste0(toCreate, folderName, "/")
+        if (!labkey.webdav.pathExists(baseUrl, folderPath, fileSet, toCreate)) {
+            if (!labkey.webdav.mkDir(baseUrl, folderPath, fileSet, toCreate)){
```
Comment (XingY) [inline Rlabkey/R/labkey.webdav.R]: this is not working for me since fileSet and toCreate order doesn't match defined arguments and toCreate is not a matched param

File: Rlabkey/R/labkey.webdav.R [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+    if (!overwrite & file.exists(localFilePath)) {
+        return()
+    }
+
+    localDownloadDir <- dirname(localFilePath)
+    if (!file.exists(localDownloadDir)) {
+        dir.create(localDownloadDir, recursive = TRUE)
+    }
+
+    options <- labkey.getRequestOptions(method="GET")
```
[the remark this answers is not in the document]
Comment (bbimber) [reply Rlabkey/R/labkey.webdav.R]: I considered that, but will labkey.get() support the write_disk() part associated with downloading a file?  This internal webdav.getByUrl() method has the checks associated with overwrite=T  as well.  

I am all for reusing code; however, I dont think what you've highlighted is quite that.  right below this lines 50-57 might be, though I believe I concluded I'd need to add some new arguments to labkey.get and b/c I was already sharing the hard part (labkey.getRequestOptions), it didnt seem worth it.

File: Rlabkey/R/labkey.webdav.R (hunk 2 above)
Comment (bbimber) [reply Rlabkey/R/labkey.webdav.R]: well that's what I originally had, but @labkey-klum suggested the re-order, given it has a default value.  I tend to agree w/ Karl; however, i dont love how R deals with arguments in general.

File: Rlabkey/R/labkey.webdav.R (hunk 3 above)
Comment (bbimber) [reply Rlabkey/R/labkey.webdav.R]: yes, i should have seen that.  the result of switching argument order and not using named arguments.  i'm gonna switch everything to use named args within labkey.webdav

File: Rlabkey/R/labkey.webdav.R [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+    return(is.null(ret$exception))
+}
+
+labkey.webdav.listDir <- function(baseUrl=NULL, folderPath, remoteFilePath, fileSet="@files")
+{
+    baseUrl=labkey.getBaseUrl(baseUrl);
+
+    url <- labkey.webdav._validateRemoteUrl(baseUrl = baseUrl, folderPath = folderPath, fileSet = fileSet, remoteFilePath = remoteFilePath)
+    url <- paste0(url, "?method=JSON")
+    content <- labkey.post(url, pbody = "", responseType = "text/plain; charset=utf-8")
```
[the remark this answers is not in the document]
Comment (bbimber) [reply Rlabkey/R/labkey.webdav.R]: Isnt accessing private methods exactly what you highlighted here?  What are you thinking I'd do differently?

File: Rlabkey/R/labkey.webdav.R [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+{
+    if (missing(localFile)) {
+        stop (paste("A value must be specified for localFile"))
+    }
+
+    if (!file.exists(localFile)){
+        stop (paste0("File does not exist: ", localFile));
+    }
+
+    url <- labkey.webdav._validateRemoteUrl(baseUrl = baseUrl, folderPath = folderPath, fileSet = fileSet, remoteFilePath = remoteFilePath)
```
[the remark this answers is not in the document]
Comment (bbimber) [reply Rlabkey/R/labkey.webdav.R]: yes, I'll fix this and add tricky characters to the test case

File: Rlabkey/R/labkey.webdav.R (hunk 5 above)
[the remark this answers is not in the document]
Comment (XingY) [reply Rlabkey/R/labkey.webdav.R]: Sry for not being clear. This is regarding the comment in the original ticket "either a) making the generic labkey.get() and .post() methods public, or b) consider a generic webdav get/put helper function?" 
Apparently it's decided on approach b). I'm just commenting that you can still use labkey.get() and post() without making the changes to Rlabkey package.

File: Rlabkey/R/labkey.webdav.R (hunk 4 above)
[the remark this answers is not in the document]
Comment (XingY) [reply Rlabkey/R/labkey.webdav.R]: You are right since you need to pass in special params can't easily reuse labkey.get here.

File: Rlabkey/R/labkey.webdav.R (hunk 5 above)
[the remark this answers is not in the document]
Comment (bbimber) [reply Rlabkey/R/labkey.webdav.R]: I dont believe labkey.get() and post() are accessible outside the Rlabkey package though?

File: Rlabkey/R/labkey.webdav.R (hunk 6 above)
[the remark this answers is not in the document]
Comment (bbimber) [reply Rlabkey/R/labkey.webdav.R]: Ok, this is now added.  The test file now includes characters needing encoding as well

File: Rlabkey/R/labkey.webdav.R (hunk 5 above)
[the remark this answers is not in the document]
Comment (XingY) [reply Rlabkey/R/labkey.webdav.R]: You can access it using Rlabkey:::labkey.get, not ideal but works...

[the remark this answers is not in the document]
Comment (bbimber) [reply Rlabkey/R/labkey.webdav.R]: No, I dont think this is right.  Code within Rlabkey can, yes, as this PR is in fact doing in several spots.  You are highlighting code that is using labkey.post.  Code outside Rlabkey cannot, since those are private methods, not exposed outside the package.  If you try to use Rlabkey::labkey.get outside this package you should get an error along the lines of "Error: 'labkey.get' is not an exported object from 'namespace:Rlabkey'".

The original question behind this PR was 'can someone write code that leverages Rlabkey to interact with LabKey servers over webdav'. While in theory someone could just make HTTP requests against a LK server using public R packages, Rlabkey has code to build up the proper headers for interacting with LabKey servers, and there is no point in re-creating that.  Therefore, I saw two ways to approach this: a) expose more basic Rlabkey methods outside the package, like labkey.get(), post(), or getRequestOptions().  The latter is really the most important one.  One potential solution would be to alter the NAMESPACE to export these methods.  b) actually write labkey.webdav methods into Rlabkey.  As you can see, I went with the latter.  With a couple exceptions, these are using the core labkey.post and .get methods.  You did highlight a couple spots where I didnt use labkey.get(), which is b/c capturing the response as a file is different than a simple get, and it still is using getRequestOptions(), which seems like the most important code not to replicate.

[the remark this answers is not in the document]
Comment (XingY) [reply Rlabkey/R/labkey.webdav.R]: I see where the confusion is. Rlabkey::labkey.get won't be accessible but Rlabkey:::labkey.get (with 3 of ':'s) should be accessible. 

Even though we could have gone with the option of directly using those private rlabkey methods, I agree the webdav addition to Rlabkey is great and very useful. I'll pull down the changes and do a round of testing.

[the remark this answers is not in the document]
Comment (bbimber) [reply Rlabkey/R/labkey.webdav.R]: Interesting.  That is strange R would allow that.

**Comment by bbimber** [comment]
Thanks - do you want me to remove webDavTest.R?  It doesnt do much in this package; however, I was going to propose using it as the basis of an automated test in SVN.

**Comment by labkey-klum** [comment]
Hi Ben, go ahead and leave the webDavTest.R, we have a small story on the board to wrap up any manual and automated testing before we merge this PR and submit to CRAN.  Trey will most likely look at your existing test code to figure out the best way to integrate it.

**Comment by bbimber** [comment]
OK, that works for me.  I'm happy to write testing since I'm the one pushing the feature, but certainly wouldnt say no if you guys want to do it.

It would perhaps be possible to piece out the API calls from webdavTest into the scheme you use for the existing RlabkeyTest and rlabkey-api.xml; however, I'd note that the nature of webdav is a bit different than query API calls in that for the latter the response is critical to test (since that's the whole point of the API), while for webdav the presence/absence of files after the API call is what matters.

Do you have a rough estimate on the timing of when this might be pushed to CRAN?

**Comment by labkey-klum** [comment]
Looking at the team's kanban board, I would anticipate that the story will get pulled on sometime next week. Assuming the team can wrap up the testing in 2-4 days maybe another week? There is also another change for a client that we would try to bundle in the same submission but it is a small task.

**Comment by labkey-tchad** [comment]
@bbimber Finally getting a chance to run this through its paces. I'm going to do some manual testing and add some regression tests to `RlabkeyTest` in trunk. I'll be adding any issues I find as comments to this pull request.

**Comment by labkey-tchad** [comment]
All of the positive scenarios are looking good. There isn't much in the way of error handling or feedback when an operation fails.
* `labkey.webdav.get`
  * Returns `TRUE` when `overwrite=FALSE` and the file already exists locally
  * Throws an error when `overwrite=TRUE` but a directory with the same name already exists locally
  * Returns `TRUE` when remote file doesn't exist
  * Creates a local file when the remote file doesn't exist. It contains the error response:
    * `{"exception":"Not Found","success":false,"status":404}`
* `labkey.webdav.mkDir`
  * Returns `TRUE` when it fails because the parent directory doesn't exist

File: Rlabkey/DESCRIPTION [hunk 7]
```diff
@@ -1,18 +1,18 @@
...
+    into the R environment for analysis, provided users have permissions
+    to read the data. It also enables R users to insert, update, and
+    delete records stored on a 'LabKey' Server, provided they have appropriate
+    permissions to do so.
+License: Apache License 2.0
+Copyright: Copyright (c) 2010-2018 LabKey Corporation
+LazyLoad: true
+Depends: httr, jsonlite
+LinkingTo: Rcpp
+Imports: Rcpp (>= 0.11.0)
```
Comment (labkey-tchad) [inline Rlabkey/DESCRIPTION]: Please revert the changes to the `DESCRIPTION` file. It looks like you changed the line endings and Cory can update this when he's ready to push to CRAN.

Comment (bbimber) [reply Rlabkey/DESCRIPTION]: Hi Trey, 

Sure, but it would help to get a sense on the automated testing plan.  For example, when I fix those scenarios, my default would be to add test cases to that existing webDavTest.R file; however, I dont know that this is how tests should look.  As I noted above, since checking webDav operations should generally involving inspecting the results of the API call and checking files, the r-labkey-api.xml scheme may not be quite right either.  Any further thoughts there?

Comment (labkey-tchad) [reply Rlabkey/DESCRIPTION]: I split up `r-labkey-api.xml` and there's now a `rlabkey-api-webdav.xml`. The test case that invokes it is currently disabled but if you comment out the `@Ignore` annotation from `RlabkeyTest.testRlabkeyWebDavApi` you'll be able to run the new webdav test cases: `./gradlew uiTest -Ptest=RlabkeyTest.testRlabkeyWebDavApi`
I think they cover all of the scenarios I mentioned and the failing test cases have the response I thought seemed to make sense.

Comment (bbimber) [reply Rlabkey/DESCRIPTION]: So I believe the basic problem is that R thinks the invalid webdav calls are succeeding.  for example:

response <- GET('https://prime-seq.ohsu.edu/_webdav/home/@files/fakeFile.txt')

reports a status of 200.  the body of the response has status_code = 404; however Rlabkey's processResponse doesnt pick up on that.  I'll investigate a little more.

Comment (bbimber) [reply Rlabkey/DESCRIPTION]: To illustrate more:
```

library(Rlabkey)
response <- GET('https://labkey.org/_webdav/fakeFile.txt')
#the response reports 200:
response$status_code 
#this is the actual JSON, which reports 404:
content(response)$status  

```
IF the header reported an error status, Rlabkey::processResponse would parse this and throw.  Because it doesnt, this looks like a successful request and that's not happening.

I am inclined to think either a) the server is doing the wrong thing (a more invasive change), or b) Rlabkey::processResponse() should be changed to cover this scenario, c) labkey.webdav should wrap processResponse() with a webdav-specific one that covers this scenario.

If anyone who knows more about HTTP requests has thoughts I would appreciate ideas.

**Comment by bbimber** [comment]
@cnathe In webdav.R, I created processWebDavResponse() as an example.   That inspects the JSON of the response for a status code, and preferentially uses this.  That fixes most of the problems Trey found.

However, generic labkey.post() requests will hit this same problem.  For example (any server could be substituted):

```
labkey.setDebugMode(T)
Rlabkey:::labkey.post('https://labkey.org/_webdav/home/@files/fakeFile.txt?method=DELETE', pbody = '')

```
This appears to R as though status=200, and therefore succeeded.  We could integrate the logic I created in  processWebDavResponse to processResponse; however, my concern is that we're missing a more general solution (i.e. httr should handle this but we're not using it right).  

have you seen anything like this before?

File: Rlabkey/R/labkey.defaults.R [hunk 8]
```diff
@@ -209,19 +207,52 @@ labkey.setDebugMode <- function(debug=FALSE)
...
     message <- error$exception
   }
-  if (haltOnError)
+  if (haltOnError) {
+    status_code <- getStatusCode(response)
+
+    # Note: is this request was writing to a file, the error message JSON will be written to that file, so we delete.  
+    if (inherits(response$content, 'path')) {
+      if (file.exists(response$content)){
+        unlink(response$content)
```
Comment (labkey-tchad) [inline Rlabkey/R/labkey.defaults.R]: Will this delete an existing file when `overwrite=TRUE` and the request has an error?

Comment (bbimber) [reply Rlabkey/R/labkey.defaults.R]: I suppose, but it would already have been overwritten with the JSON content of this response.

The basic problem is that our WebDav API returns 200 status codes for many types of failures, and then includes the true status/exception in the JSON body.  As this currently works in R/httr, when I make these requests we're asking R to write the response to a file.  Under normal operation, that response is the actual desired file.  We dont generally want to write this whole file to memory.  For the code you point out to be called, that request would have already happened, and the result of the response written to that file.  

Also - I believe what's checked in fixes all your requests except 2:

- If labkey.get() is made against a directory, the server returns the HTML/JS code of the FileBrowser.  That's just how the WebDav controller works.  I am trying to think of ways around this, but dont have one right now.

- I'll have a fix in a minute; however, your issue with downloadFolder() has to do w/ not handling spaces in the folderPath properly, I think.  That's why it worked as expected for me and my example (webDavTest.R), but not yours.  I should have a fix on this soon.

Comment (bbimber) [reply Rlabkey/R/labkey.defaults.R]: @labkey-tchad - I think the current version of this should fix the problems you noted.

File: Rlabkey/R/labkey.webdav.R [hunk 9]
```diff
@@ -0,0 +1,265 @@
...
+    response <- labkey.webdav.getByUrl(url, localFilePath, overwrite)
+
+    return(file.exists(localFilePath))
+}
+
+labkey.webdav.getByUrl <- function(url, localFilePath, overwrite=TRUE)
+{
+    # dont bother querying if this file already exists, since we wont overwrite it
+    if (!overwrite & file.exists(localFilePath)) {
+        return()
```
Comment (labkey-tchad) [inline Rlabkey/R/labkey.webdav.R]: Could this return an indication that it no-opped?
```suggestion
        return(FALSE)
```

Comment (bbimber) [reply Rlabkey/R/labkey.webdav.R]: OK.  I had to change a little more to make this pass through and I also updated the man page to document that behavior.

**Comment by bbimber** [comment]
hi trey, yes, i think you're right and contextPath is the problem.  i have been doing all my testing against a server w/o one.  i'll fix this.

i'm not sure i agree on throw/warn.  first, while file/dir warn, HTTP requests in R tend to error.  i dont think we're inherently tied to file-conventions.  this is an HTTP request.  plus, all the other R APIs are going to tend to error, not warn.  why should labeky.webdav behave differently than labkey.selectrows?  They each share response-handling code in labkey.defaults, and I'd be reluctant to make them have different conventions.  in the situations you outline, it seems like it could be argued those scenarios are actually errors or incorrect use of the API.

second, i personally find it really annoying how difficult it can be to make R die after error.   a warning is even that much more obscured, especially when run as part of a larger piece of code.  

i could imagine situations where expanding the throwOnError flag makes sense.  it there were cases where issuing some call not knowing if it will succeed makes sense, perhaps we expand the methods that support this?

**Comment by bbimber** [comment]
hi trey - my last commit adds a dependency on urltools, but should properly handle contextPath parsing.

**Comment by labkey-tchad** [comment]
A couple more issues:
1. R doesn't like that `if` statement at the end of `webdav.get`. Getting this error when downloading:
`Error in if (ret == FALSE) { : missing value where TRUE/FALSE needed`

1. `webdav.downloadFolder` is still creating too many folders. The entire `remoteFilePath` is created in the `localDir`. `(remoteFilePath='a/b/c' localDir='download')` creates `download/a/b/c` instead of `download/c`.

Regarding errors vs warnings: that seems reasonable; I'll leave the API design to you all.

Could we handle the context path without pulling in a new dependency? LabKey webdav URLs always start with `/_webdav`; why not just trim off everything before that?

**Comment by labkey-tchad** [comment]
A question about `downloadFolder`. It looks like the `overwrite` flag only applies to actual files, not directories and will merge the downloaded folder with existing folders. Is that the intended behavior?

**Comment by bbimber** [comment]
Sorry, i finally get what you've been saying on downloadFolder.  I'll check that and add a test case.  

adding urltools doesnt seem like that big a deal, but I can rework this to not include the dependency.  i think you're right about checking for /_webdav.

can you say a little about when you hit that problem with get()?

**Comment by bbimber** [comment]
HI Trey, regarding downloadFolder() and overwrite=T: I hadnt thought a lot about that.  I'm trying to imagine usage, and I would argue the current behavior makes sense, or at least we should preserve some kind of flag to keep that.  for example, a common use case would be to mirror a remote directory locally.  I could see runing this with override=F so we dont re-download files, but i would want new files to get downloaded, which might not happen if we stop when we encounter an existing folder.  I guess I could see a second flag governing directory merging if you felt this was important.

**Comment by bbimber** [comment]
@labkey-tchad :

I believe I fixed downloadDirectory to stop including the extra directories.

I was not able to repro the "Error in if (ret == FALSE)" error for get(): can you say more about when you hit this?

**Comment by labkey-tchad** [comment]
`downloadFolder` can't handle spaces:
```
Error in curl::curl_fetch_memory(url, handle = handle) : 
 Host name 'RlabkeyTest Project' contains bad letter
Calls: labkey.webdav.downloadFolder ... request_fetch -> request_fetch.write_memory -> <Anonymous> -> .Call
```
I changed that if statement to use `ifTRUE` and it's working for me now:
```
if (isTRUE(ret == FALSE)) {
      return(FALSE)
    }
```
I can email you some debug output if you'd like to investigate further

**Comment by bbimber** [comment]
Hi trey, something must be different b/c i dont repro either of those things.  The fact that your error is complaining about hostName, but returning the folderPath suggests either something is calling it weirdly or something in the code is parsing it badly?

I'd also like to figure out this isTRUE() thing.  The scenario when I could see this mattering would be if ret was NULL or NA (in which case ret == FALSE is NA, isTRUE(NA) is FALSE).  By wrapping this with isTRUE() we avoid an error but we might be masking bad behavior.  

Are your repro cases for these part of the checked in RLabkey test?

**Comment by labkey-tchad** [comment]
Yes. The test also logs the full script now for any failures so that you can repro manually after a failure.

**Comment by bbimber** [comment]
hey,

two questions:

1) where'd the test code go?  is there a different git repo?

2) if your tests log the full script, would it be possible to just post the failing cases?  i dont have a full trunk instance going right now.

**Comment by labkey-tchad** [comment]
The Rlabkey tests are in [SVN](https://svn.mgt.labkey.host/stedi/trunk/server/test/data/api/rlabkey-api-webdav.xml). The Selenium portion of the test does a bunch of setup; it creates some files and directories for testing and does a bunch of string replacement in the R scripts. It isn't easy to run the tests outside the test harness.

This is the `webdav.get` script errors for me:
```
labkey.webdav.get(baseUrl="http://localhost:8080/labkey", folderPath="RlabkeyTest Project", remoteFilePath="remote/readChecks/getMe.txt", localFilePath="/home/tchad/development/labkey/trunk/build/deploy/files/RlabkeyTest Project/@files/webdav_download/getMe.txt", overwrite=TRUE)
```
The `webdav.downloadFolder` is similar:
```
labkey.webdav.downloadFolder(baseUrl="http://localhost:8080/labkey", folderPath="RlabkeyTest Project", remoteFilePath="remote/readChecks/getMe", localDir="/home/tchad/development/labkey/trunk/build/deploy/files/RlabkeyTest Project/@files/webdav_download", overwrite=TRUE)
```

**Comment by bbimber** [comment]
@labkey-tchad : OK, so the downloadFolder issue was a simple typo I accidentally added.  now fixed.

I still dont repro you issues w/ get(). I added another check for is.null on the response though.

**Comment by labkey-tchad** [comment]
The file I'm attempting to download in that test is just 5 random bytes. I wonder if that's causing some problem

**Comment by labkey-tchad** [comment]
This is the end of the debug log when I try the `get` on the nonsense file:
```
[1] "URL: http://localhost:8080/labkey/_webdav/RlabkeyTest%20Project/@files/remote/readChecks/getMe.txt"
*  Found bundle for host localhost: 0x5561751b26f0 [can pipeline]
*  Re-using existing connection! (#0) with host localhost
*  Connected to localhost (127.0.0.1) port 8080 (#0)
*  Server auth using Basic with user 'tchad@labkey.com'
-> GET /labkey/_webdav/RlabkeyTest%20Project/@files/remote/readChecks/getMe.txt HTTP/1.1
-> Host: localhost:8080
-> Authorization: Basic dGNoYWRAbGFia2V5LmNvbTp5ZWtiYWwxIQ==
-> User-Agent: Rlabkey
-> Accept-Encoding: gzip, deflate
-> Cookie: JSESSIONID=E78DFC76634DE213A6FA2FCC3C57CFB2; X-LABKEY-CSRF=bd2827c2f8616ef5308530db79c62775
-> Accept: application/json, text/xml, application/xml, */*
-> 
<- HTTP/1.1 200 
<- X-Frame-Options: SAMEORIGIN
<- X-XSS-Protection: 1; mode=block
<- X-Content-Type-Options: nosniff
<- Referrer-Policy: origin-when-cross-origin
<- ETag: W/"5-1557522415135"
<- Last-Modified: Fri, 10 May 2019 21:06:55 GMT
<- Expires: 
<- Cache-Control: private
<- Pragma: private
<- Vary: Cookie
<- Content-Type: text/plain;charset=UTF-8
<- Content-Length: 5
<- Date: Fri, 10 May 2019 21:19:45 GMT
<- 
<<  ��r�

*  Connection #0 to host localhost left intact
Error in if (ret == FALSE) { : missing value where TRUE/FALSE needed
```

**Comment by bbimber** [comment]
@labkey-tchad It is possible we have different versions of R or relevant packages?  I added one more null check.  my theory is still that some configuration of packages/files results in content() returning some type of null value.  If that's correct, I dont think isTRUE() would do the right thing.  Does your nonsense file actually get downloaded?  If so, then returning FALSE in this scenario is probably incorrect, right?  if I'm wrong and your file isnt getting downloaded, I will check more.  

note: i just added the extra null check this morning.

**Comment by labkey-tchad** [comment]
The nonsense file does get downloaded. Your latest changed passes all of the `webdav.get` test cases in `RlabkeyTest`. Not quite sure what was getting returned that was choking us there.

`webdav.downloadFolder` is leaving out the entire remote file path now. Is that the expected behavior? Maybe this could work similar to `mv` and `cp` shell commands. If `localDir` already exists, we download the remote folder there, including the remote folder's name. If `localDir` doesn't exist, we create it (`dir.create(localDir, recursive=FALSE)`) and download the contents of the remote folder into it.

Also, I don't think we came to a conclusion regarding behavior of the `overwrite` flag for `downloadFolder`. I propose adding the `merge` flag that you mentioned rather than forcing assumptions onto a single parameter:
* `merge && overwrite`: Download all files from the remote folder, replacing any existing files with downloaded ones.
* `merge && !overwrite`: Download all files from the remote folder as long as they don't exist already, leaving existing files.
* `!merge && overwrite`: [**default**] Blow away the local directory and replace it with the contents of the remote directory.
* `!merge && !overwrite`: Return `FALSE` if local directory already exists and is not empty; otherwise, download remote directory.

We may want to consider doing the initial download to a temp directory so that we don't end up with an incomplete download if something fails part way through.

**Comment by bbimber** [comment]
OK, glad the first one is resolved.

On webdav.downloadFolder: yes, I was assuming we would keep it relatively simple and drop the path (this seemed to be what your earlier posts suggested, but I do get your point here about cp/mv).  It seems that with the current behavior, the person using the API could still accomplish this?  Examples:

Remote: /foo/bar:
Local: /myLocalDir/bar
Result: downloads the contents of the remote folder into /myLocalDir/bar

Remote: /foo/bar:
Local: /myLocalDir/
Result: downloads the contents of the remote folder into /myLocalDir/, which does not end up with the folder named "bar"

The user would need to request the proper download location.  If we support automatically creating folders in the manner you suggest, we also need to think about what happens if the parent doesnt exist (i.e. i think the internal logic needs to be more complicated than 'does target exist, if not create').  I would therefore be inclined to keep as-is and just document the expected behavior.  Another way to think about this is that the downloads behave like requesting "/remoteDir/*".  If you fell strongly I can look into it.

On the merge/overwrite piece: my intent was to try to keep the APIs simple.  I wrote downloadFolder as a wrapper since this seemed like a pretty common use-case.  I am a little reluctant to make this too complicated, but let me see what I can do about a merge flag.  I would note the current behavior acts as though merge=T, which I still think isnt crazy.  We're not trying to re-create rsync here.

**Comment by labkey-tchad** [comment]
I've updated the test to match the current behavior. Seeing as I don't have to use or maintain this library, I'll leave it to you, Karl, and Cory to finalize any design decisions.

**Comment by bbimber** [comment]
@labkey-tchad hey - sorry if there was a miscommunication.  i just added the mergeFolder behavior you outlined above.  Since I was working on it, I also added the mv/cp esque behavior you wanted:

If localDir already exists, we download the remote folder there, including the remote folder's name. If localDir doesn't exist, we create it.

I updated the man page to reflect both of these.  Note: to avoid confusion on downloadFolder, I renamed 'overwrite' to 'overwriteFiles', and added 'mergeFolders' to handle the merge behavior you outlined.  If I did this right, it should follow the four combos you outline above.

**Comment by labkey-tchad** [comment]
Updated the test but `downloadFolder` isn't behaving as you describe. Its still splatting the folder contents into `localDir`. It will create `localDir` if it doesn't exist but it won't create a directory matching the name of the remote directory automatically. This also makes the `mergeFolders` flag confusing as it only applies after the root folder.

**Comment by bbimber** [comment]
Thinking about this though, does that run into problems if someone runs the command multiple times, excepting to merge?  for example:

Round 1:
downloadFolder('/localDir', 'remoteDir')
Result: local dir doesnt exist, so we create it, then download the contents of remoteDir into it.

Round 2: repeat same command:
Result: local dir now exists, so instead of deterministically downloading into the same location, we would take the basename of remoteDir, make a subfolder, and load there?  I dont like this.

Second, I see you point on the arguably odd behavior of mergeFolder.  I'm a little concerned about a high potential for unwanted deleting on mergeFolders=F, especially if that will be the default.  the consequence of getting overwriteFiles incorrect is that you lose one one, and that file probably existed remotely.  if you do the wrong thing with mergeFolders, and we're going to delete whole directories, bad things could happen easily.  that was why I wrote it to never delete the top-level.  That would make more sense if we always made one more level with the directory name and loaded into that though.  

Perhaps these would argue for:  treating "localFolder" more like "localBaseDir", and always creating a subfolder with the remote directory name?  I'll need to think this through, but I wonder if this would give us consistency across all the scenarios.

**Comment by bbimber** [comment]
Hi Trey - I'm hopefully maybe this can satisfy each of our concerns.  downloadFolder() will always create a subfolder with the basename of the remoteDir.  Note: the argument is called localBaseDir, not localDir.  This means we can give the behavior you've been asking for, and we can safely delete is mergeFolder=F.

**Comment by labkey-tchad** [comment]
Please update `rlabkey-api-webdav.xml` on trunk and verify that `downloadFolder` works as expected.
I just realized I didn't add a check for how `downloadFolder` handles a 404. Please add a test for that as well.
When you're finished Cory will review and commit a build of `2.2.6` to trunk then push to CRAN after it's had a couple of days to bake on TeamCity.

**Comment by bbimber** [comment]
@labkey-tchad hi trey - i have the Rlabkey webdav tests passing and committed the updates to trunk.  i needed to update RLabkey test to normalize filepaths so they pass on windows as well.

**Comment by cnathe** [comment]
@bbimber I pulled down the latest code for this PR and build locally. Attached are the warning messages from the CRAN check command. Ben, would you mind fixing up these doc related warnings?
[00check.log](https://github.com/LabKey/labkey-api-r/files/3202797/00check.log)

**Comment by bbimber** [comment]
@cnathe - i just committed changes.  R CMD check is now clean for me.

**Comment by cnathe** [comment]
@bbimber Thanks, Ben. I pulled those changes, built, and ran the CRAN check again locally and it looks clean for me as well.

@labkey-tchad The last things before merging this PR is to review the TeamCity results. Trey, it looks like there were some RlabkeyTest failures overnight. I don't think they are directly related to these changes but would you mind reviewing the test results and see what you think? Thanks.

**Comment by labkey-tchad** [comment]
The Rlabkey tests are failing on Windows but I'm not sure why.
```
Error: package or namespace load failed for 'Rlabkey' in rbind(info, getNamespaceInfo(env, "S3methods")):
 number of columns of matrices must match (see arg 2)
```

**Comment by bbimber** [comment]
I did all dev on a windows machine, so in theory this would work there.  do the tests work off a one-time generated tar.gz file?  if so, it seems like perhaps something may be bad about this one file?  do windows and linux tests each use the same artifact for tests?

**Comment by bbimber** [comment]
FWIW, if you google that error: "rbind(info, getNamespaceInfo(env, "S3methods")):", you'll find a variety of people reporting this kind of thing.  there was no obvious reason it would pop up on windows agents here though.

**Comment by labkey-tchad** [comment]
Still failing. I wonder if there's some backward compatibility issue. The Windows agents are running 3.5.2 but Rlabkey was built with 3.6

**Comment by bbimber** [comment]
@labkey-tchad i have R 3.5.6 - want me to try something with an artifact I built?

**Comment by cnathe** [comment]
@labkey-tchad @bbimber I had built the Rlabkey 2.3.0 version on my Windows machine with R version 3.6.0 since CRAN has rejected a package submission in the past if it wasn't built with the latest version. I will rebuilt the Rlabkey 2.3.0 version with R 3.5.2 locally and check that in to see if TeamCity is happier with that one. If so, then I guess we will see what CRAN says.

**Comment by bbimber** [comment]
@labkey-tchad is there a way to see more detailed information on TeamCity from the execution of these scripts?  It might exist and I'm just not seeing it.  I only see the build log, which contains the failure of assertEquals and not much detail about the actual text from executing R.

I did a remote run using artifacts I built with R 3.5.3; however, they also failed DailyB SQLServer:

https://teamcity.labkey.org/viewLog.html?buildTypeId=LabkeyTrunk_DailySqlserverb&buildId=753539&tab=testsInfo&pager.currentPage=1

I am probably just missing it, but I'm not finding much useful information to understand the failure.

**Comment by labkey-tchad** [comment]
The test log includes the actual output, the expected text, and the full input script. They're hard to pick out though because they're not very visually distinct:
```
07:05:38,064       Error: could not find expected text: [1] "Latest Rlabkey version: 2.3.0".
from value:
javax.script.ScriptException: javax.script.ScriptException: An error occurred when running the script 'script.R', exit code: 1).
Loading required package: httr
Loading required package: jsonlite
Error: package or namespace load failed for 'Rlabkey' in rbind(info, getNamespaceInfo(env, "S3methods")):
 number of columns of matrices must match (see arg 2)
In addition: Warning messages:
1: package 'Rlabkey' was built under R version 3.6.0 
2: package 'httr' was built under R version 3.5.3 
3: package 'jsonlite' was built under R version 3.5.3 
Execution halted
```
```
07:05:38,064     Expected results for test case: Rlabkey version check:
[1] "Latest Rlabkey version: 2.3.0"
[1] "Found Rlabkey version: 2.3.0"
```
```
07:05:38,064     Script for failed test case: Rlabkey version check:
.libPaths("D:/teamcity/work/e2019250b39c72d6/sampledata/rlabkey")
library(Rlabkey)
print("Latest Rlabkey version: 2.3.0")
print(paste("Found Rlabkey version:", installed.packages()["Rlabkey","Version"]))
```

**Comment by labkey-tchad** [comment]
Your remote run didn't work because we don't actually use the RLabkey in source control at that point. There's a separate build that bundles up all of the R packages.

**Comment by cnathe** [comment]
@labkey-tchad @bbimber Success! The TeamCity run with the last Rlabkey v2.3.0 package looks good. I will be merging this PR shortly.

**Comment by bbimber** [comment]
Excellent - thank you all for helping get this added.  @cnathe - do you have any guess how long the CPAN submission process takes?

**Comment by cnathe** [comment]
@bbimber I got an automated email from CRAN that the process is underway. It usually takes a couple of days before the bits for the new version make it up to CRAN (in my past experience anyway).
